tools/c7n_azure/tests/azure_common.py

```diff
@@ -12,115 +12,306 @@
 
         return r1_path == r2_path
 
-    def request_callback(self, request):
+    def _request_callback(self, request):
         """Modify requests before saving"""
-        if "/subscriptions/" in request.url:
-            request.uri = re.sub(
-                r"[\da-zA-Z]{8}-([\da-zA-Z]{4}-){3}[\da-zA-Z]{12}",
-                DEFAULT_SUBSCRIPTION_ID,
-                request.url)
+        request.uri = AzureVCRBaseTest._replace_subscription_id(request.uri)
+        request.uri = AzureVCRBaseTest._replace_tenant_id(request.uri)
+
         if request.body:
             request.body = b'mock_body'
+
+        # Request headers serve no purpose as only URI is read during a playback.
+        request.headers = None
+
         if re.match('https://login.microsoftonline.com/([^/]+)/oauth2/token', request.uri):
             return None
         if re.match('https://login.microsoftonline.com/([^/]+)/oauth2/token', request.uri):
             return None
         return request
 
+    def _response_callback(self, response):
+        if self.is_playback():
+            if 'data' in response['body']:
+                body = json.dumps(response['body']['data'])
+                response['body']['string'] = body.encode('utf-8')
+                response['headers']['content-length'] = [str(len(body))]
+
+            return response
+
+        headers = {}
+        for key in response['headers']:
+            if key.lower() not in self.FILTERED_HEADERS:
+                headers[key.lower()] = response['headers'][key]
+        response['headers'] = headers
+
+        content_type = response['headers'].get('content-type', (None,))[0]
+        if not content_type or 'application/json' not in content_type:
+            return response
+
+        body = response['body'].pop('string').decode('utf-8')
+
+        # Clean up subscription IDs and storage keys
+        body = AzureVCRBaseTest._replace_subscription_id(body)
+        body = AzureVCRBaseTest._replace_storage_keys(body)
+
+        try:
+            response['body']['data'] = json.loads(body)
+        except json.decoder.JSONDecodeError:
+            self.fail("AzureVCRBaseTest could not parse JSON response body "
+                      "while attempting to record cassette. Body:\n%s" % body)
+
+        # Replace some API responses entirely
+        response = AzureVCRBaseTest._response_substitutions(response)
+
+        return response
+
+    @staticmethod
+    def _response_substitutions(response):
+        data = response['body']['data']
+
+        if isinstance(data, dict):
+            odata_metadata = data.get('odata.metadata')
+            if odata_metadata and "directoryObjects" in odata_metadata:
+                response['body']['data'] = GRAPH_RESPONSE
+                return response
+
+            # Replace Activity Log API responses
+            value_array = data.get('value', [])
+            if len(value_array) > 0 and value_array[0].get('eventTimestamp'):
```

Comment:
(minor) `if len(value_array) > 0` can be replaced with `if value_array`.

appears to be recommended in https://www.python.org/dev/peps/pep-0008/

Reply:
Yes of course, not sure why but I had in my head that the [] in this case was differentiated from None.